### ISSUE TITLE
indexed_fill: sharding correctness/validation bugs (WIDTH_SHARDED batch-offset, overly strict shard-match validation, COL_MAJOR misaddressing)

### ISSUE THREAD

**Issue by abdulla-TT** [tenstorrent/tt-metal#55673]
### Describe the bug

Found while doing the TM Layout / Sharding Support Audit for #25529: [TM Layout / Sharding Support Audit](https://docs.google.com/document/d/1shGKatKb5HM-xNM6v07OR4K1uVcFkmb0dIYSsGO7lGE/edit?tab=t.0#heading=h.e9k48ka0sxmy) (see the "Per-op sharding support matrix" section, `indexed_fill` row).

`ttnn.indexed_fill` has several correctness/validation bugs in its sharding support:

1. **Silent wrong data on multi-row `WIDTH_SHARDED` grids.** The shard-local kernel computed `batch_offset_a = shard_row * total_batches_per_core` and `cx = i % shard_n_x` for every shard-local core. That's only correct for `BLOCK_SHARDED`. For `WIDTH_SHARDED`, every core must see the full batch range and its column index is just its linear position `i`. On any `WIDTH_SHARDED` grid with more than one row (e.g. 2x2, 2x4), cores outside row 0 silently dropped replacements (reproduced at PCC ≈ 0.75-0.87).

2. **Overly strict shard-match validation.** `WIDTH_SHARDED`/`BLOCK_SHARDED` input hard-rejected any output that didn't exactly match its sharding (e.g. `INTERLEAVED`), even though `HEIGHT_SHARDED` input already falls back correctly to the generic path for a non-matching output. Valid generic-path fallback configurations (e.g. `dim != 0`, DRAM sharding, unsupported sharded input_b) were also incorrectly rejected by the same check.

3. **`COL_MAJOR`-oriented shards misaddressed by the shard-local kernel.** The shard-local kernel derives each core's shard/column index from its row-major position in the core enumeration, which only matches the tensor's actual shard-to-core assignment when the shard orientation is `ROW_MAJOR`. A `COL_MAJOR`-oriented `WIDTH_SHARDED`/`BLOCK_SHARDED` shard on a multi-row/column grid would read/write the wrong shard, silently producing incorrect data.

4. **Validation and program-factory path selection used different output configs**, so a caller passing an explicit shard-spec-less `MemoryConfig` could hit validation preconditions for the shard-local kernel even though the factory would actually select the generic path (or vice versa).

Additionally, `indexed_fill` had no test coverage at all for DRAM-sharded or ND-sharded inputs.

### To reproduce

Example for bug 1 (silent wrong data on a multi-row `WIDTH_SHARDED` grid):

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

B, H, W, b = 4, 64, 128, 2
core_grid = ttnn.CoreGrid(y=2, x=2)  # multi-row grid triggers the bug

batch_id = torch.randint(0, B, (1, 1, 1, b))
batch_id_ttnn = ttnn.Tensor(batch_id, ttnn.uint32).to(
    device, ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.L1)
)

width_sharded_mem_config = ttnn.create_sharded_memory_config(
    shape=(B, 1, H, W), core_grid=core_grid, strategy=ttnn.ShardStrategy.WIDTH
)

torch_a = torch.rand((B, 1, H, W), dtype=torch.bfloat16)
torch_b = torch.rand((b, 1, H, W), dtype=torch.bfloat16)
input_tensor_a = ttnn.from_torch(
    torch_a, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT, memory_config=width_sharded_mem_config
)
input_tensor_b = ttnn.from_torch(
    torch_b, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT,
    memory_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.L1),
)

output_tensor = ttnn.indexed_fill(
    batch_id_ttnn, input_tensor_a, input_tensor_b, memory_config=width_sharded_mem_config
)
# Replacements on grid rows > 0 are silently dropped; PCC vs. golden ≈ 0.75-0.87 instead of ~1.0.
```

### Expected behavior

`indexed_fill` should produce numerically correct results for all supported sharding layouts (`HEIGHT_SHARDED`, `WIDTH_SHARDED`, `BLOCK_SHARDED`, DRAM-sharded, ND-sharded) and should fall back to the generic layout-agnostic path (rather than hard-reject) whenever the shard-local/native fast-path preconditions aren't met.

### Fix

Fixed in #55620:
- Corrected `batch_offset_a`/`cx` computation for `WIDTH_SHARDED` in the shard-local program factory.
- Relaxed `validate_on_program_cache_miss` to only enforce shard-local preconditions when the shard-local kernel will actually be selected (using the same eligibility check the program factory uses).
- Added a `ROW_MAJOR`-orientation requirement to the shard-local eligibility check, falling back to the generic path for `COL_MAJOR` shards.
- Made program-factory path selection use the same (resolved) output memory config as validation.
- Added regression/coverage tests for multi-row `WIDTH_SHARDED`, sharded-in/interleaved-out, DRAM-sharded, ND-sharded, and `COL_MAJOR` orientation fallback.
